sklearn/neighbors/lmnn.py

```diff
@@ -0,0 +1,843 @@
+from ..utils.fixes import argpartition
+from ..utils.multiclass import check_classification_targets
+from ..utils.validation import check_is_fitted, check_array, check_X_y, \
+    check_random_state
+from ..exceptions import DataDimensionalityWarning
+
+
+class LargeMarginNearestNeighbor(KNeighborsClassifier):
+    """Distance Metric Learning for Large Margin Classification
+
+    Large margin nearest neighbor classification (LMNN) is a machine learning
+    algorithm for metric learning. It learns a (pseudo-)metric in a
+    supervised fashion to improve the classification accuracy of the k-nearest
+    neighbor rule.
+    The main intuition behind LMNN is to learn a pseudometric under which all
+    data instances in the training set are surrounded by at least k instances
+    that share the same class label. If this is achieved, the leave-one-out
+    error is minimized.
+    This implementation follows closely Kilian Weinberger's MATLAB code found
+    at <https://bitbucket.org/mlcircus/lmnn> which solves the unconstrained
+    problem, finding a linear transformation with L-BFGS instead of solving the
+    constrained problem that finds the globally optimal metric.
+
+
+    Parameters
+    ----------
+    L : array-like
+        Initial transformation in an array with shape (n_features_out,
+        n_features_in).  If None `load` will be used to load a transformation
+        from a file. (default: None)
+
+    n_neighbors : int
+        Number of target neighbors (default: 3)
+
+    max_iter : int
+        Maximum number of iterations in the optimization (default: 200)
+
+    use_pca : bool
+        Whether to use pca to warm-start the linear transformation.
+        If False, the identity will be used. (default: True)
+
+    tol : float
+        Tolerance for the optimization  (default: 1e-5)
+
+    n_features_out : int
+        Preferred dimensionality of the inputs after the transformation.
+        If None it is inferred from `use_pca` and `L`.(default: None)
+
+    max_constr : int
+        Maximum number of constraints to enforce per iteration (default: 10
+        million).
+
+    use_sparse : bool
+        Whether to use a sparse or a dense matrix for the impostor-pairs
+        storage. Using a sparse matrix, the distance to impostors is computed
+        twice, but it is somewhat faster for larger data sets than using a
+        dense matrix. With a dense matrix, the unique impostor pairs have to be
+        identified explicitly (default: True).
+
+    load : string
+        A file path from which to load a linear transformation. If None, either
+        identity or pca will be used based on `use_pca` (default: None).
+
+    save : string
+        A file path prefix to save intermediate linear transformations to.
+        After every function call, it will be extended with the function call
+        number and the `.npy` file extension. If None, nothing will be saved
+        (default: None).
+
+    disp : int, optional
```

Comment:
I'm considering having two booleans as options then. One (verbose) that prints information from my gradient computing function and one (iprint) if the user wants to get messages from the optimizer. I have observed that the messages from this optimizer are not printed so consistently. Usually there is a bulk of messages printed together every 50 iterations, while a user would most likely want to get a progress report in every iteration. I guess this is because it is fortran code and printing to stdout does not work as expected with the python port.